tortoise/ninja_tortoise.go

```diff
@@ -65,81 +66,116 @@ type Mesh interface {
 
-	tGood map[types.LayerID]votingPattern //good pattern for layer i
+	TGood map[types.LayerID]votingPattern //good pattern for layer i
 
-	tVote map[votingPattern]map[types.BlockID]vec //global opinion
+	TVote map[votingPattern]map[types.BlockID]vec //global opinion
 }
 
-func NewNinjaTortoise(layerSize int, blocks Mesh, hdist int, log log.Log) *ninjaTortoise {
-	return &ninjaTortoise{
-		Log:          log,
-		Mesh:         blocks,
-		hdist:        types.LayerID(hdist),
-		avgLayerSize: layerSize,
-		pBase:        ZeroPattern,
+func NewNinjaTortoise(layerSize int, blocks Mesh, hdist int, log log.Log) *NinjaTortoise {
 
-		patterns:   map[types.LayerID][]votingPattern{},
-		tGood:      map[types.LayerID]votingPattern{},
-		tEffective: map[types.BlockID]votingPattern{},
-		tCorrect:   map[types.BlockID]map[types.BlockID]vec{},
-		tExplicit:  map[types.BlockID]map[types.LayerID]votingPattern{},
-		tSupport:   map[votingPattern]int{},
-		tPattern:   map[votingPattern]map[types.BlockID]struct{}{},
+	trtl := &NinjaTortoise{Log: log, Mesh: blocks}
+
+	trtl.ninjaTortoise = &ninjaTortoise{
+		Hdist:        types.LayerID(hdist),
+		AvgLayerSize: layerSize,
+		PBase:        ZeroPattern,
+
+		Patterns:   map[types.LayerID][]votingPattern{},
+		TGood:      map[types.LayerID]votingPattern{},
+		TEffective: map[types.BlockID]votingPattern{},
+		TCorrect:   map[types.BlockID]map[types.BlockID]vec{},
+		TExplicit:  map[types.BlockID]map[types.LayerID]votingPattern{},
+		TSupport:   map[votingPattern]int{},
+		TPattern:   map[votingPattern]map[types.BlockID]struct{}{},
+
+		TVote:              map[votingPattern]map[types.BlockID]vec{},
+		TTally:             map[votingPattern]map[types.BlockID]vec{},
+		TComplete:          map[votingPattern]struct{}{},
+		TEffectiveToBlocks: map[votingPattern][]types.BlockID{},
+		TPatSupport:        map[votingPattern]map[types.LayerID]votingPattern{},
+	}
+
+	return trtl
+}
+
+func (ni *NinjaTortoise) saveOpinion() error {
+	for bid, vec := range ni.TVote[ni.PBase] {
+		valid := vec == Support
+		if err := ni.SaveContextualValidity(bid, valid); err != nil {
+			return err
+		}
+
+		if !valid {
+			ni.With().Warning("block is contextually invalid", log.BlockId(bid.String()))
+		}
+		events.Publish(events.ValidBlock{Id: bid.String(), Valid: valid})
+	}
+	return nil
+}
 
-		tVote:              map[votingPattern]map[types.BlockID]vec{},
-		tTally:             map[votingPattern]map[types.BlockID]vec{},
-		tComplete:          map[votingPattern]struct{}{},
-		tEffectiveToBlocks: map[votingPattern][]types.BlockID{},
-		tPatSupport:        map[votingPattern]map[types.LayerID]votingPattern{},
+func (ni *NinjaTortoise) PersistTortoise() error {
+	if err := ni.saveOpinion(); err != nil {
+		return err
 	}
+	return ni.Persist(mesh.TORTOISE, ni.ninjaTortoise)
 }
 
-func (ni *ninjaTortoise) evictOutOfPbase() {
+func (ni *NinjaTortoise) RecoverTortoise() (interface{}, error) {
+	return ni.Retrieve(mesh.TORTOISE, &ninjaTortoise{})
```

Comment:
we're saving the whole tortoise in one key?

Reply:
yes we dont have a need for any granularity, partial data has no use right now